Upgraded Vibration Analysis Module

jpbarraca opened the pull request:
An attempt to address #610

Currently the values are sampled at the FC and sent to GCS in chunks of 16 values per axis.
Spectrogram was enhanced in order to process this data and show it. Optionally GCS can calculate an FFT using the FFTReal library.

Improvements:
- Can potentially higher number of bins (1024 tested)
- Memory consumption is reduced at the FC
- Analysis parameters can be modified on runtime without restarting GCS
- Spectrogram can apply the FFT or not (optional)
- Enable the application of further analysis from the RAW accel values
- Reduced load in the FC

Drawbacks:
- Telemetry bandwidth is increased and may lead to a bursty behaviour.
- Adds an external header (FFTReal with Free License)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/822)

<!-- Reviewable:end -->

jpbarraca committed:
Upgraded Vibration Analysis Module

mlyle commented:
This looks really good!

It's getting late in the Tanto cycle, so I don't consider this a candidate for merging for the next release.  But this fixes most of my complaints about vibration analysis.

The main thing I see-- I'm not sure we should be converting back and forth from Q15 instead of just keeping floats.  What do you think?

@droidicus -- this now uses a very similar strategy to your new-autotune approach.

jpbarraca commented:
I left the Q15 conversions because I do not know if native floats can create problems in other boards (soft float?).
If you know it is safe I can clean it.

jpbarraca committed:
Fixed signed/unsigned comparison

mlyle commented:
Yes-- all the data comes in as floats, all of our F3 and F4 parts have a FPU (and F1 doesn't have memory or code space for this).  So while it'll use slightly more memory, I think avoiding the conversion to fixed point is a good idea.

mluessi commented on flight/Modules/VibrationAnalysis/vibrationanalysis.c:
```diff
@@ -86,8 +84,6 @@ static struct VibrationAnalysis_data {
 	int16_t *accel_buffer_complex_x_q15;
```
The data is no longer complex, so the variables should be renamed.

mluessi commented on ground/gcs/src/plugins/scope/scopes3d/ffft/DynArray.h:
```diff
@@ -0,0 +1,101 @@
+the extent permitted by applicable law. You can redistribute it
+and/or modify it under the terms of the Do What The Fuck You Want
+To Public License, Version 2, as published by Sam Hocevar. See
+http://sam.zoy.org/wtfpl/COPYING for more details.
```
Is this license GPL compatible? If not, we could use [FFTW](http://www.fftw.org/)

jpbarraca replied:
Don't know about hidden consequences of this license, but this seems to be very compatible:
...
TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
1. You just DO WHAT THE FUCK YOU WANT TO.

jpbarraca commented:
@mluessi I'm cleaning up the code as we speak. ~~Handling floats directly and~~ removing traces of complex math.

mluessi commented:
Regarding Q15/float, I wonder if it would be better to quantize the data to 16bits when sending it to the GCS via UAVO. After all, the sensor is 16bits, so we shouldn't lose much and it would reduce the data rate by 50%.

jpbarraca commented:
@mluessi Given the purpose of this module, that may be acceptable. One drawback of using floats and 16 value chunks is that the UAVObject is rather large and will introduce jitter to other telemetry data.

jpbarraca committed:
Kept Q15 format in order to reduce memory usage.
Values transmitted in Q15 format in order to reduce telemetry bandwidth.
Scale value (currently constant, but can be adaptive to minimize error) is sent with each UAVO instance.
Minor refactoring.

mlyle commented:
If we do fixed point 16-bit, it probably shouldn't be Q15; we set up the sensor to be + / - 16G usually, but Q15 can measure only 0-1G; it means Z will spend most of the time clipped and relatively small vibrations can cause clipping in other axes (which in turn will generate odd/even harmonics).

jpbarraca committed:
Apply custom scale from float to int16 to minimize error.

jpbarraca commented:
Well... there are a few bugs and is not ready for prime time. Will provide a fixed version for the next milestone.

mlyle commented:
OK.  I appreciate it, whenever you're ready :)
